st2api/st2api/controllers/v1/auth.py

```diff
@@ -0,0 +1,156 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pecan
+import six
+
+from pecan import abort
+from pecan.rest import RestController
+from mongoengine import ValidationError
+
+from st2common import log as logging
+from st2common.models.api.auth import ApiKeyAPI
+from st2common.models.api.base import jsexpose
+from st2common.exceptions.auth import ApiKeyNotFoundError
+from st2common.persistence.auth import ApiKey
+from st2common.rbac.types import PermissionType
+from st2common.rbac.decorators import request_user_has_permission
+from st2common.rbac.decorators import request_user_has_resource_permission
+from st2common.util import auth as auth_util
+
+http_client = six.moves.http_client
+
+LOG = logging.getLogger(__name__)
+
+__all__ = [
+    'ApiKeyController'
+]
+
+
+class ApiKeyController(RestController):
+    """
+    Implements the REST endpoint for managing the key value store.
+    """
+
+    def __init__(self):
+        super(ApiKeyController, self).__init__()
+        self.get_one_db_method = ApiKey.get_by_key_or_id
+
+    @request_user_has_resource_permission(permission_type=PermissionType.API_KEY_VIEW)
+    @jsexpose(arg_types=[str])
+    def get_one(self, api_key_id_or_key):
+        """
+            List api keys.
+
+            Handle:
+                GET /apikeys/1
+        """
+        api_key_db = None
+        try:
+            api_key_db = ApiKey.get_by_key_or_id(api_key_id_or_key)
+        except ApiKeyNotFoundError:
+            msg = 'ApiKey matching %s for reference and id not found.', api_key_id_or_key
+            LOG.exception(msg)
+            abort(http_client.NOT_FOUND, msg)
+
+        try:
+            return ApiKeyAPI.from_model(api_key_db)
+        except (ValidationError, ValueError) as e:
+            LOG.exception('Failed to serialize API key.')
+            abort(http_client.INTERNAL_SERVER_ERROR, str(e))
+
+    @request_user_has_permission(permission_type=PermissionType.API_KEY_VIEW)
```

Comment:
The RBAC scoping for this is still not sorted so I have gone with a slightly relaxed primitive i.e. you get to view if you have permission on the object. This means an admin can view keys for every user etc.

I suppose this is "sorta" like an admin being able to see another users password.

I do like the you only get to see it once at generation idea and did not really consider the approach. Whenever I've used API keys I often tend to go back to the application to grab the old keys so it would make me slightly annoyed to generate a new one. Having said that I have no strong opinion about this behavior and am happy to experiment.